lib/Transforms/IPO/StructFieldCacheAnalysisImpl.h

```diff
@@ -0,0 +1,377 @@
+
+  /// Increment stats for one category
+  void addStats(unsigned Category) { StatCounts[Category]++; }
+  /// Print a brief stats of struct access
+  void printStats();
+
+private:
+  const Module &CurrentModule;
+
+  /// Function reference that is used to retrive execution count for basic block
+  function_ref<BlockFrequencyInfo *(Function &)> LookupBFI;
+
+  /// A map storing access info of all structs
+  std::unordered_map<const StructType *, StructFieldAccessInfo *>
+      StructFieldAccessInfoMap;
+
+  StructHotnessAnalyzer *HotnessAnalyzer;
+
+  /// \name Data structure to get statistics of each DebugStats entry
+  /// %{
+  std::vector<unsigned> StatCounts;
+  const std::vector<std::string> StatNames = {
+      "Variable type is Struct**",
+      "Function argument is a value",
+      "Function argument is not defined in the program",
+      "GEP value passed into indirect function calls or function that has "
+      "undetermined num args",
+      "GEP value passed into bitcast",
+      "GEP value passed into unexpected opcode",
+      "User is not Instruction nor Operator",
+      "Struct defined but no accesses",
+      "Struct passed into functions defined out of scope",
+      "GEP instruction directly used on struct*",
+      "Unknown instruction directly used on struct*",
+      "Struct filtered out due to colder than a ratio of maximum hotness"};
+  /// %}
+
+  /// Used to print name of each StructDefinitionType
+  const std::vector<std::string> StructDefinitionTypeNames = {
+      "global struct", "global struct*", "global struct**",
+      "local struct",  "local struct*",  "local struct**"};
+};
+
+/// This class is used to store all access information for each struct
+/// declared in the program. It records all loads and stores to all fields
+/// of the struct to provide essential information for cache-aware struct
+/// field analysis.
+class StructFieldAccessInfo {
+private:
+  // This struct organizes a call on a function with each argument access which
+  // struct field
+  struct FunctionCallInfo {
+    FunctionCallInfo(const Function *F, unsigned ArgNum, FieldNumType FieldNum)
+        : FunctionDeclaration(F) {
+      Arguments.resize(FunctionDeclaration->arg_size());
+      assert(ArgNum < Arguments.size());
+      Arguments[ArgNum] = FieldNum;
+    }
+    void insertCallInfo(unsigned ArgNum, FieldNumType FieldNum) {
+      assert(ArgNum < Arguments.size());
+      Arguments[ArgNum] = FieldNum;
+    }
+    const Function *FunctionDeclaration;
+    FieldNumArrayType Arguments;
+  };
+  // This struct organizes all calls on a function definition with all mappings
+  // of arguments and struct field number
+  struct FunctionAccessPattern {
+    FunctionAccessPattern(FieldNumArrayType *CallSite) {
+      CallSites.clear();
+      CallSites.push_back(CallSite);
+    }
+    void insertCallInfo(FieldNumArrayType *CallSite) {
+      CallSites.push_back(CallSite);
+    }
+    std::vector<FieldNumArrayType *> CallSites;
```

Comment:
I plan to switch to LLVM containers in later version (SmallVectors and DenseMaps) as you suggested. I'd like to keep using STL containers just for safe for now.